### PR TITLE
Issue #3245: Only request database updates if needed

### LINKED ISSUE #3245: [UX] Hide "Run Database Updates" / next steps if there are no next steps
## Describe your issue or idea
When updating a module via the UI, there is not always a database update to complete, but the "Next Steps" section always shows with the "Run database updates" link:

screen shot 2018-08-07 at 10 52 11 am

When you click the link, you are informed that there are no database updates pending (and a warning about backing up before you run updates). 

screen shot 2018-08-07 at 10 51 54 am

It would be better to just display a "No updates pending" message after the first step.

### PR DESCRIPTION
**Pull request by laryn**
FIxes https://github.com/backdrop/backdrop-issues/issues/3245

### CODE REVIEW

**Comment by laryn** [comment]
@klonos Would you be willing to test this one? It seems like it should work but at the end of the process the module isn't updated, it seems. :confused:

**Comment by yorkshire-pudding** [comment]
@laryn - I have tested this with the patch on a local installation and can confirm the behaviour you described in the [comment above](https://github.com/backdrop/backdrop/pull/3847#issuecomment-984839542) - it goes through the process and looks good but doesn't update the version.
chrome_bbEzKXgpr5

Tested the same action (manual installation of latest minus 1 release of `flag`) on a test site without the patch and getting the same result so don't think that this PR is causing that issue.

@klonos - please can you see if you get the same result?

**Comment by yorkshire-pudding** [comment]
@laryn @klonos - I think this is due to https://github.com/backdrop/backdrop-issues/issues/3997
I don't think it should block this PR as most people won't manually install an older version, and even if they do, they will experience the [issue 3997](https://github.com/backdrop/backdrop-issues/issues/3997) issue regardless of this PR.

This is a significant improvement to UX so would be good to get into the milestone.  I have reviewed the code and it all looks good.

**Review by yorkshire-pudding — APPROVED** [review]
I have reviewed the code and think it all looks good

**Comment by yorkshire-pudding** [comment]
@laryn @klonos - I've reviewed the code and am happy.  Does it still also need a code review from Greg, or should that be removed?

**Comment by herbdool** [comment]
@yorkshire-pudding though you were able to confirm that the other bug is not related it this PR, it's not clear that you tested it fully. Were you able to confirm:

1. Using installer to update a module that does *not* have updates will not require running update.php
2. Using installer to update a module that *does* have updates will still require running update.php.

**Comment by yorkshire-pudding** [comment]
@herbdool 
Was able to confirm 1.
Took a while but finally found a suitable test case.  Installed first version of paragraphs on a site without patch and had db updates.
Did same on site with patch and doesn't ask for updates.

**Comment by herbdool** [comment]
Thanks for testing @yorkshire-pudding. I guess it needs a bit more work.

File: core/modules/installer/installer.authorize.inc
Comment (yorkshire-pudding) [inline core/modules/installer/installer.authorize.inc:234]: Testing this for `paragraphs` in `devel` 's Execute PHP.  Once the code update is done and the list of updates is shown on `core/update.php` - this function doesn't return anything.

File: core/modules/installer/installer.authorize.inc
Comment (yorkshire-pudding) [inline core/modules/installer/installer.authorize.inc:238]: Once `core/update.php` was invoked, the `current_schema` was 999 so if a `latest_schema` was returned it would work.  However, before this was invoked, it was showing as 7102.  Now, most update with be in the 1000 range and only initial ones might be 7000+ but other functions have more complex comparisons.

Comment (argiepiano) [reply core/modules/installer/installer.authorize.inc:238]: The issue with running  `backdrop_get_schema_version($module)` at this point is that most of the `.install` files have not been included by PHP. And of course, all of the hook_update_N() functions live in the `.install` files. So yes, this will always fail to detect needed update hooks.

Comment (yorkshire-pudding) [reply core/modules/installer/installer.authorize.inc:238]: I had seemed to get results fairly consistently from `backdrop_get_installed_schema_version($module)` ; with `paragraphs`, I got what was in the database (i.e. `7102`) until I had invoked `core/update.php` and then the function in `paragraphs.install` kicked in and provided the value of `999`.

However, I never got anything from `backdrop_get_schema_versions($module)`.

I think this will need someone who understands these functions to have a look and think about how to get the information out at this stage.

**Pull request closed without merging**
